experimental/lib/Analysis/GsaAnalysis.cpp

```diff
@@ -0,0 +1,485 @@
+      std::vector<std::string> conditionsList;
+
+      // Vector associating each input of the Phi to a boolean expression
+      std::vector<std::pair<boolean::BoolExpression *, GateInput *>>
+          expressionsList;
+
+      // For each input of the phi, compute the boolean expression which defines
+      // its usage
+      for (auto &operand : phiOperands) {
+
+        // Remove the current operand from the list of blocks to avoid
+        blocksToAvoid.erase(blocksToAvoid.begin());
+
+        // Find all the paths from `commonDominator` to `phiBlock` which pass
+        // through operand's block but not through any of the `blocksToAvoid`
+        auto paths = findAllPaths(commonDominator, phiBlock,
+                                  operand->getBlock(), blocksToAvoid);
+
+        boolean::BoolExpression *phiInputCondition =
+            boolean::BoolExpression::boolZero();
+
+        // Sum all the conditions for each path
+        for (auto &path : paths) {
+          auto *condition = getPathExpression(path, conditionsList);
+          phiInputCondition =
+              boolean::BoolExpression::boolOr(condition, phiInputCondition);
+          phiInputCondition = phiInputCondition->boolMinimize();
+        }
+
+        // Associate the expression to the phi operand
+        expressionsList.emplace_back(phiInputCondition, operand);
+      }
+
+      // Sort the conditions according to their index
+      llvm::sort(conditionsList.begin(), conditionsList.end(),
+                 [](std::string a, std::string b) {
+                   a.erase(0, 1);
+                   b.erase(0, 1);
+                   return std::stoi(a) < std::stoi(b);
+                 });
+
+      // Expand the expressions to get the tree of gammas
+      auto *gammaRoot = expandGammaTree(expressionsList, conditionsList, phi);
+      gammaRoot->isRoot = true;
+
+      // Once that a phi has been converted into a tree of gammas, all the
+      // gates which used the original phi as input must be connected to the
+      // root of the gamma tree
+      for (auto &[bb, phis] : gateList) {
+        for (auto &phii : phis) {
+          for (auto &op : phii->operands) {
+            if (op->isTypeGate() && op->getGate() == phi)
+              op->input = gammaRoot;
+          }
+        }
+      }
+    }
+  }
+}
+
+template <typename FunctionType>
+void GsaAnalysis<FunctionType>::convertPhiToMu(FunctionType &funcOp) {
+
+  if (funcOp.getBlocks().size() == 1)
+    return;
+
+  mlir::DominanceInfo domInfo;
+  mlir::CFGLoopInfo loopInfo(domInfo.getDomTree(&funcOp.getBody()));
+
+  // For each phi
+  for (auto const &[phiBlock, phis] : gateList) {
+    for (auto *phi : phis) {
+
+      // A phi might be a MU iff it is inside a for loop and has exactly
+      // two operands
+      if (!loopInfo.getLoopFor(phiBlock) || phi->operands.size() != 2)
+        continue;
+
+      auto *op0Block = phi->operands[0]->getBlock(),
```

Comment:
Same as above, don't overuse `auto` when the type is simple/small enough to write. See [rationale](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable).